File: Backend/ClearChain.API/Services/PickupRequestSummary.cs
using ClearChain.Domain.Entities;

namespace ClearChain.API.Services;

/// <summary>
/// A pickup request's line items are the source of truth for what it moves. The
/// denormalized <c>Listing*</c> columns on <see cref="PickupRequest"/> are only a summary
/// of those items, kept so a client that has the request but not its items (an offline
/// cache, a push notification payload) still has something truthful to show.
///
/// Every write path must call <see cref="Apply"/> after building the items so the summary
/// can never drift from them.
/// </summary>
public static class PickupRequestSummary
{
    public const string MultipleCategory = "Multiple";
    public const string MixedUnit = "items";

    /// <summary>Recomputes the request's summary columns from its items.</summary>
    public static void Apply(PickupRequest request)
    {
        var items = request.Items.ToList();
        if (items.Count == 0)
            return;

        request.ListingTitle = Title(items);
        request.ListingCategory = Category(items);
        request.ListingUnit = Unit(items);
        request.ListingExpiryDate = EarliestExpiry(items);
        request.RequestedQuantity = items.Sum(i => i.RequestedQuantity);
    }

    /// <summary>
    /// One item reads as the product itself; several read as a count, because no single
    /// product name can stand for the rest.
    /// </summary>
    public static string Title(IReadOnlyCollection<PickupRequestItem> items) =>
        items.Count switch
        {
            0 => string.Empty,
            1 => items.First().ListingTitle,
            _ => $"{items.Count} items"
        };

    public static string Category(IReadOnlyCollection<PickupRequestItem> items)
    {
        var distinct = items
            .Select(i => i.ListingCategory)
            .Where(c => !string.IsNullOrWhiteSpace(c))
            .Distinct(StringComparer.OrdinalIgnoreCase)
            .ToList();

        return distinct.Count == 1 ? distinct[0] : MultipleCategory;
    }

    public static string Unit(IReadOnlyCollection<PickupRequestItem> items)
    {
        var distinct = items
            .Select(i => i.ListingUnit)
            .Where(u => !string.IsNullOrWhiteSpace(u))
            .Distinct(StringComparer.OrdinalIgnoreCase)
            .ToList();

        return distinct.Count == 1 ? distinct[0] : MixedUnit;
    }

    /// <summary>The whole request is only as fresh as its soonest-expiring item.</summary>
    public static string? EarliestExpiry(IReadOnlyCollection<PickupRequestItem> items) =>
        items
            .Select(i => i.ListingExpiryDate)
            .Where(d => !string.IsNullOrWhiteSpace(d))
            .OrderBy(d => d, StringComparer.Ordinal)
            .FirstOrDefault();
}

File: Mobile/app/src/main/java/com/clearchain/app/domain/model/PickupRequestSummary.kt
package com.clearchain.app.domain.model

/**
 * A pickup request's [PickupRequest.items] are the source of truth for what it moves; the
 * `listing*` fields on the request itself are only a summary the server derives from them
 * (see PickupRequestSummary on the API side).
 *
 * These helpers are the one place that knows how to read a request's contents, so screens
 * can't each invent their own rule.
 */

/**
 * The request's contents as displayable lines.
 *
 * Falls back to a single line rebuilt from the summary because the Room cache stores
 * requests without their items, so an offline list would otherwise show nothing.
 */
val PickupRequest.itemLines: List<PickupRequestItem>
    get() = items.ifEmpty {
        listOf(
            PickupRequestItem(
                id = id,
                requestedQuantity = requestedQuantity,
                listingTitle = listingTitle,
                listingCategory = listingCategory,
                listingExpiryDate = listingExpiryDate,
                listingUnit = listingUnit
            )
        )
    }

/**
 * Everything a person might type to find this request, lowercased. Searching only the
 * summary title would miss the individual products inside a multi-item request.
 */
val PickupRequest.searchText: String
    get() = buildString {
        append(listingTitle).append(' ')
        append(listingCategory).append(' ')
        append(groceryName).append(' ')
        append(ngoName).append(' ')
        notes?.let { append(it).append(' ') }
        items.forEach { item ->
            append(item.listingTitle).append(' ')
            append(item.listingCategory).append(' ')
        }
    }.lowercase()

/** Product names in the request, for exports and summaries. */
val PickupRequest.itemTitles: List<String>
    get() = itemLines.map { it.listingTitle }.filter { it.isNotBlank() }
